core-rust/state-manager/src/store/parent_map.rs

```diff
@@ -0,0 +1,315 @@
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use crate::store::traits::*;
+use crate::{ChangeAction, SubstateChange};
+
+use radix_engine::types::*;
+
+/// A resolver of [`SubstateNodeMetadata`] for [`SubstateChange`]s.
+pub struct NodeMetadataResolver {
+    /// A map of created Nodes to their parent Substates, extracted from the [`SubstateChange`]s.
+    changed_parent_by_node: NonIterMap<NodeId, SubstateReference>,
+    /// A map of existing Nodes to their root Nodes (i.e. `SubstateNodeMetadata#root`), loaded from
+    /// a [`SubstateNodeMetadataStore`] for each Node that does not have a parent in the
+    /// [`changed_parent_by_node`] map.
+    existing_root_by_node: NonIterMap<NodeId, SubstateReference>,
+}
+
+impl NodeMetadataResolver {
+    /// Resolves the [`SubstateNodeMetadata`] for all newly-created Nodes found in the [`Create`]d
+    /// and [`Update`]d substates. (within the given [`SubstateChange`]s).
+    /// The resolution considers the trees (or rather: potentially-incomplete bottom-up tree
+    /// fragments) constructed within the given [`SubstateChange`]s, and the top fragments of these
+    /// trees already-existing in the given [`SubstateNodeMetadataStore`].
+    /// API note: the results are grouped by [`SubstateNodeMetadata`] (since it may be shared by
+    /// many [`NodeId`]s).
+    pub fn batch_resolve<S: SubstateNodeMetadataStore>(
```

Comment:
Nice. Didn't quite realise how intrinsically complicated this resolution is, but it looks correct to me, and can't see any obvious simplifications :D